File: LottoPlayer/app/src/main/java/com/im/daeseong/lottoplayer/Util/RoundView.java
package com.im.daeseong.lottoplayer.Util;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.view.View;

public class RoundView extends View {

    private static int DEFAULT_TITLE_COLOR = Color.WHITE;
    private static int DEFAULT_BACKGROUND_COLOR = Color.CYAN;
    private static final int DEFAULT_VIEW_SIZE = 80;
    private static float DEFAULT_TITLE_SIZE = 25f;
    private static String DEFAULT_TITLE = "0";

    private int mTitleColor = DEFAULT_TITLE_COLOR;
    private int mBackgroundColor = DEFAULT_BACKGROUND_COLOR;
    private String mTitleText = DEFAULT_TITLE;
    private float mTitleSize = DEFAULT_TITLE_SIZE;

    private TextPaint mTitleTextPaint;
    private Paint mBackgroundPaint;
    private RectF mInnerRectF;
    private int mViewSize;

    private Typeface mFont = Typeface.defaultFromStyle(Typeface.NORMAL);

    public RoundView(Context context) {
        super(context);
        init(null, 0);
    }

    public RoundView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0);
    }

    public RoundView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(attrs, defStyle);
    }

    private void init(AttributeSet attrs, int defStyle) {

        //Title TextPaint
        mTitleTextPaint = new TextPaint();
        mTitleTextPaint.setFlags(Paint.ANTI_ALIAS_FLAG);
        mTitleTextPaint.setTypeface(mFont);
        mTitleTextPaint.setTextAlign(Paint.Align.CENTER);
        mTitleTextPaint.setLinearText(true);
        mTitleTextPaint.setColor(mTitleColor);
        mTitleTextPaint.setTextSize(mTitleSize);

        //Background Paint
        mBackgroundPaint = new Paint();
        mBackgroundPaint.setFlags(Paint.ANTI_ALIAS_FLAG);
        mBackgroundPaint.setStyle(Paint.Style.FILL);
        mBackgroundPaint.setColor(mBackgroundColor);

        mInnerRectF = new RectF();
    }

    private void invalidateTextPaints(){
        mTitleTextPaint.setTypeface(mFont);
        mTitleTextPaint.setTextSize(mTitleSize);
        mTitleTextPaint.setColor(mTitleColor);
    }

    private void invalidatePaints(){
        mBackgroundPaint.setColor(mBackgroundColor);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        int width = resolveSize(DEFAULT_VIEW_SIZE, widthMeasureSpec);
        int height = resolveSize(DEFAULT_VIEW_SIZE, heightMeasureSpec);
        mViewSize = Math.min(width, height);

        setMeasuredDimension(width, height);
    }

    @Override
    protected void onDraw(Canvas canvas) {

        mInnerRectF.set(0, 0, mViewSize, mViewSize);
        mInnerRectF.offset((getWidth() - mViewSize) / 2, (getHeight() - mViewSize) / 2);

        float centerX = mInnerRectF.centerX();
        float centerY = mInnerRectF.centerY();

        int xPos = (int) centerX;
        int yPos = (int) (centerY - (mTitleTextPaint.descent() + mTitleTextPaint.ascent()) / 2);

        canvas.drawOval(mInnerRectF, mBackgroundPaint);

        canvas.drawText(mTitleText, xPos, yPos, mTitleTextPaint);
    }

    public String getTitleText() {
        return mTitleText;
    }

    public void setTitleText(String title) {
        mTitleText = title;
        invalidate();
    }

    public int getBackgroundColor() {
        return mBackgroundColor;
    }

    public void setBackgroundColor(int backgroundColor) {
        mBackgroundColor = backgroundColor;
        invalidatePaints();
    }

    public float getTitleSize() {
        return mTitleSize;
    }

    public void setTitleSize(float titleSize) {
        mTitleSize = titleSize;
        invalidateTextPaints();
    }

    public void setTextTypeface(Typeface font){
        this.mFont = font;
        invalidateTextPaints();
    }

    public void setTitleColor(int titleColor) {
        mTitleColor = titleColor;
        invalidateTextPaints();
    }
}

File: LottoPlayer(kotlin)/app/src/main/java/com/daeseong/lottoplayer/Util/RoundView.kt
package com.daeseong.lottoplayer.Util

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Typeface
import android.text.TextPaint
import android.util.AttributeSet
import android.view.View

open class RoundView : View {

    companion object {
        private const val DEFAULT_TITLE_COLOR: Int = android.graphics.Color.WHITE
        private const val DEFAULT_BACKGROUND_COLOR: Int = android.graphics.Color.CYAN
        private const val DEFAULT_VIEW_SIZE = 80
        private const val DEFAULT_TITLE_SIZE = 25f
        private const val DEFAULT_TITLE = "0"
    }

    private var mTitleColor = DEFAULT_TITLE_COLOR
    private var mBackgroundColor = DEFAULT_BACKGROUND_COLOR
    private var mTitleText = DEFAULT_TITLE
    private var mTitleSize = DEFAULT_TITLE_SIZE

    private var mTitleTextPaint: TextPaint? = null
    private var mBackgroundPaint: Paint? = null
    private var mInnerRectF: RectF? = null
    private var mViewSize = 0

    private var mFont = Typeface.defaultFromStyle(Typeface.NORMAL)

    constructor(context: Context?) : super(context) {
        init(null, 0)
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init(attrs, 0)
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle) {
        init(attrs, defStyle)
    }

    private fun init(attrs: AttributeSet?, defStyle: Int) {

        // Title TextPaint
        mTitleTextPaint = TextPaint().apply {
            flags = Paint.ANTI_ALIAS_FLAG
            typeface = mFont
            textAlign = Paint.Align.CENTER
            isLinearText = true
            color = mTitleColor
            textSize = mTitleSize
        }

        // Background Paint
        mBackgroundPaint = Paint().apply {
            flags = Paint.ANTI_ALIAS_FLAG
            style = Paint.Style.FILL
            color = mBackgroundColor
        }

        mInnerRectF = RectF()
    }

    open fun invalidateTextPaints() {
        mTitleTextPaint?.apply {
            typeface = mFont
            textSize = mTitleSize
            color = mTitleColor
        }
    }

    open fun invalidatePaints() {
        mBackgroundPaint?.color = mBackgroundColor
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val width = resolveSize(DEFAULT_VIEW_SIZE, widthMeasureSpec)
        val height = resolveSize(DEFAULT_VIEW_SIZE, heightMeasureSpec)
        mViewSize = width.coerceAtMost(height)
        setMeasuredDimension(width, height)
    }

    override fun onDraw(canvas: Canvas) {
        mInnerRectF?.apply {
            set(0f, 0f, mViewSize.toFloat(), mViewSize.toFloat())
            offset((width - mViewSize) / 2f, (height - mViewSize) / 2f)
            val centerX = centerX()
            val centerY = centerY()
            val xPos = centerX.toInt()
            val yPos = (centerY - (mTitleTextPaint!!.descent() + mTitleTextPaint!!.ascent()) / 2).toInt()
            canvas.drawOval(this, mBackgroundPaint!!)
            canvas.drawText(mTitleText, xPos.toFloat(), yPos.toFloat(), mTitleTextPaint!!)
        }
    }

    open fun getTitleText(): String? {
        return mTitleText
    }

    open fun setTitleText(title: String) {
        mTitleText = title
        invalidate()
    }

    open fun getBackgroundColor(): Int {
        return mBackgroundColor
    }

    override fun setBackgroundColor(backgroundColor: Int) {
        mBackgroundColor = backgroundColor
        invalidatePaints()
    }

    open fun getTitleSize(): Float {
        return mTitleSize
    }

    open fun setTitleSize(titleSize: Float) {
        mTitleSize = titleSize
        invalidateTextPaints()
    }

    open fun setTextTypeface(font: Typeface?) {
        mFont = font
        invalidateTextPaints()
    }

    open fun setTitleColor(titleColor: Int) {
        mTitleColor = titleColor
        invalidateTextPaints()
    }
}
